### DISCUSSION TITLE
Plot Line Color for each PlotPoint?

### DISCUSSION THREAD

**Discussion by Bastl34** [emilk/egui#3582] [Q&A]
Is it possible somehow to define a color for each `PlotPoint` when drawing a `Line`?
I want to mark specific plot points as red/yellow/green to be able to see that the plot point is "good" or not.

I'm currently using this to draw FPS for example:
```rust
let fps_points: PlotPoints = fps_chart.iter().enumerate().map(|(i, value)| {
    [
        i as f64,
        *value as f64,
        // COLOR?
    ]
}).collect();

let fps = Line::new(fps_points).color(Color32::WHITE).style(LineStyle::Solid).name("FPS");
Plot::new("FPS").show(ui, |plot_ui| {
    plot_ui.line(fps);
});
```

**Comment by tomtuamnuq** [comment]
Hi @Bastl34, I have stumbed upon your question since I want to create a ScatterPlot with points of different size and color. The `PlotItem` implementations (e.g. `Line` and `Points`) use a single style for all the elements, meaning that all parts of the line and all the points have the same color.

A [PlotPoint](https://github.com/emilk/egui/blob/d190df7d259e664a41d220f29c812da2aeb2a93e/crates/egui_plot/src/items/values.rs#L12) is just a coordinate without a color, and the color is set per Line (over all PlotPoints - see [shapes method for PlotItem trait](https://github.com/emilk/egui/blob/d190df7d259e664a41d220f29c812da2aeb2a93e/crates/egui_plot/src/items/mod.rs#L445)). The styled parts of the line (segments) use the same [color of stroke](   https://github.com/emilk/egui/blob/d190df7d259e664a41d220f29c812da2aeb2a93e/crates/egui_plot/src/items/values.rs#L97). 

A possible workaround would be to first draw the line and then draw the special points on top of it. Please see the [plot points of different radius discussion](https://github.com/emilk/egui/discussions/2214) for possible ways of drawing points of different size and color. An example:

```rust
use egui::Color32;

impl eframe::App for TemplateApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let fps_points: egui_plot::PlotPoints = [25, 24, 30, 35, 45, 41]
                .into_iter()
                .enumerate()
                .map(|(i, value)| [i as f64, value as f64])
                .collect();
            let mut circles = Vec::new();
            let egui::Response { ctx, layer_id, .. } = egui_plot::Plot::new("plot")
                .show(ui, |plt| {
                    fps_points.points().iter().for_each(|coords| {
                        if coords.y < 30.0 || coords.y > 40.0 {
                            let color = if coords.y < 30.0 {
                                Color32::RED
                            } else {
                                Color32::GREEN
                            };
                            circles.push(egui::Shape::circle_filled(
                                plt.screen_from_plot(*coords),
                                10.0, // set radius of special points to be visible
                                color,
                            ));
                        }
                    });
                    let fps = egui_plot::Line::new(fps_points)
                        .color(Color32::BLACK)
                        .width(5.0)
                        .style(egui_plot::LineStyle::Solid)
                        .name("FPS");
                    plt.line(fps);
                })
                .response;
            ctx.layer_painter(layer_id).extend(circles);
        });
    }
}
```

![SmartSelect_20240125-122538_Chrome](https://github.com/emilk/egui/assets/76523782/6e860141-6297-46a2-af77-51ef92066469)
